### PR TITLE
Handle Tensor returns in PropagateUnbackedSymInts

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #123735
* #124785
* #124782
* #124739
* #124394
* #124316
* #124314
* #124310
* __->__ #124297
* #124290

This subsumes https://github.com/pytorch/pytorch/pull/124069

In the original PR, my idea was that when we run PropagateUnbackedSymInts, we check that the sizes before and after are exactly the same. This ended up turning up lots of bugs that I didn't feel like fixing. Separately, Ivan let me know that this pass was quite expensive in terms of compile time, since we spent a lot of time thinking about the equalities.

To kill two birds with one stone, we now only check for equality precisely when an unbacked SymInt was bound (thanks to the previous PR in this stack, we now have this information). Specifically, we look to see if `meta["unbacked_bindings"]` is set on the old node, and if it is, we assert the old value is equal to the new value from the repropagation. Note that the pytree key is used to actually extract the new value from the example value, as it may be nested inside an, e.g., tensor size.

We do something a bit naughty at the end: we use `defer_runtime_assert` to actually teach ShapeEnv about the equality. This is implementationally equivalent to what we used to do, but we're going to change this later soon.

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

File: torch/fx/experimental/symbolic_shapes.py [hunk 1]
```diff
@@ -265,6 +266,29 @@ def check_consistent(new, old) -> None:
...
+    of the old and new unbacked SymInts.  Pass us the old torch.fx.Node (which
+    has the old binding information) and the new result (which we can extract the
+    new unbacked SymInts out from).
+    """
+    if bindings := n.meta.get("unbacked_bindings"):
+        for raw_u0, path in bindings.items():
+            u1 = pytree.key_get(result, path)
+            # We should never have bindings for raw bools; instead they should
+            # have been converted to ints via ConvertIntKey
+            assert type(u1) is not bool
```
Comment (bdhirsh) [inline torch/fx/experimental/symbolic_shapes.py:284]: maybe I'm holding it wrong - but shouldn't this assert be against `SymBool` (the symbolic expression stored on the FX node), not an actual bool/int/float?

File: torch/fx/experimental/symbolic_shapes.py [hunk 2]
```diff
@@ -265,6 +266,29 @@ def check_consistent(new, old) -> None:
...
+            if isinstance(u1, (int, float)):
+                raw_u1 = sympy.sympify(u1)
+            else:
+                raw_u1 = u1.node.expr
+            # TODO: replace with rename unbacked to
+            shape_env.defer_runtime_assert(sympy.Eq(raw_u0, raw_u1), "")
```
Comment (bdhirsh) [inline torch/fx/experimental/symbolic_shapes.py:290]: more of a curiosity question / me trying to understand, but: I imagined that `raw_u1` is the new expression that we want to statically "ignore", and declare equivalent to the old expression `raw_u0`. Why does that involve adding a runtime assert? (or does `defer_runtime_assert` exit early here)

File: torch/fx/experimental/symbolic_shapes.py (hunk 1 above)
Comment (ezyang) [reply torch/fx/experimental/symbolic_shapes.py:284]: You are indeed right, and I end up having to fix this code better at https://github.com/pytorch/pytorch/pull/124394/files#diff-03557db7303b8540f095b4f0d9cd2280e1f42f534f67d8695f756ec6c02d3ec7L282-R311 later in the stack (the problem doesn't manifest until the inductor tests are online though)

File: torch/fx/experimental/symbolic_shapes.py (hunk 2 above)
Comment (ezyang) [reply torch/fx/experimental/symbolic_shapes.py:290]: Yeah, this is the TODO. It ends up getting fixed in https://github.com/pytorch/pytorch/pull/124316

**Review by lezcano — APPROVED** [review]

**Pull request closed without merging**